docs/proposals/05-etcd-operator-tasks.md

```diff
@@ -0,0 +1,364 @@
+
+> Note: Dependent ordering among tasks will be addresed later which may enable concurrent execution of tasks.
+
+#### Deletion
+
+`Etcd-druid` will ensure that it garbage collects the custom resources and any other k8s resources created to realise the task as per `.spec.ttlSecondsAfterFinished` if defined in the [spec](#spec) or it assumes a default expiry time.
+
+### Use Cases
+
+#### Recovery from permanent quorum loss
+
+Currently identification and recovery from permament quorum loss is done manaually. The current proposal keeps the identification as manual, requiring an human operator to intervene and confirm that there is indeed a permanent quorum loss with no auto healing possibility. Once that is established then the next step to recover the cluster is now proposed to be automated. Recovery today is a [multi-step process](https://github.com/gardener/etcd-druid/blob/master/docs/operations/recovery-from-permanent-quorum-loss-in-etcd-cluster.md) and needs to be done carefully. Automation would ensure that we eliminate any errors from an operator.
+
+##### Task Config
+
+We do not need any config for this task. When creating an instance of `EtcdOperatorTask` for this scenario, `.spec.config` will be set to nil (unset).
+
+##### Pre-Conditions
+
+* There should be a quorum loss in a multi-member etcd cluster. For a single-member etcd cluster, invoking this task is unnecessary as the restoration of the single member is automatically handled by the backup-restore process.
+* There should not already be a permanent-quorum-loss-recovery-task running for the same etcd cluster.
+
+#### Trigger on-demand snapshot compaction
+
+`etcd-druid` provides a configurable [etcd-events-threshold](https://github.com/gardener/etcd-druid/blob/master/docs/proposals/02-snapshot-compaction.md#druid-flags). If and when this threshold is breached then only a [snapshot compaction](https://github.com/gardener/etcd-druid/blob/master/docs/proposals/02-snapshot-compaction.md) is triggered for etcd cluster. However, there are scenarios where an ad-hoc snapshot compaction is required.
+
+##### Possible scenarios
+
+* Full snapshot taken via compaction job is better as compared to full snapshot taken directly from etcd cluster as full snapshot taken via snapshot compaction is taken after [defragmentation](https://etcd.io/docs/v3.2/op-guide/maintenance/#defragmentation) which makes it smaller (more compacted) in size.
+* Full snapshot triggered on etcd cluster might cause load on running etcd cluster but full snapshot taken by compaction job is developed separately with zero effect on the running cluster.
+
+##### Task Config
+
+We do not need any config for this task. When creating an instance of `EtcdOperatorTask` for this scenario, `.spec.config` will be set to nil (unset).
+
+##### Pre-Conditions
+
+* There should not be a `on-demand snapshot-compaction` task already running for the same etcd cluster.
+
```

Comment:
Also, I can think of one more pre-condition, if the `on-demand maintenance` has been triggered some time before this task is triggered, can we reject this by having a threshold for the time before which the maintenance has been completed? As the maintenance triggers `etcdctl` compaction & defrag, then we can just do a `on-demand full` after that and can reject this `on-demand compaction`. I guess compaction is a heavy work unlike `on-demand delta/full` so I feel we can reject this whenever we can.

If this is already discussed but just not documented here, then sorry for bringing in the details in this revision :(

Reply:
Now, understand this scenario: 
Etcd db size spikes -> operator triggers the `on-demand maintenance` -> etcd db size come down -> they ping us to check why db size spiked ?
Now, to study all the events, we have our delta snapshots, now by invoking task `on-demand snapshot-compaction` with disabling defrag and compact, it will apply all the delta-snapshots on etcd cluster and can study all the db size spike ... which we already discussed `Build tool to analyse delta snapshot contents`.
So, IMO these are independent task let it be independents

Reply:
Makes sense. I Didn't quite think about the troubleshooting aspect of this. For the current revision we can go with this, but if at some point we discover some ordering of tasks, we can improve it anytime later. Thanks.